### PR TITLE
Add TypeScript support (atdts)

### PR DESCRIPTION
**Pull request by mjambon**
### PR checklist

- [x] New code has tests to catch future regressions
- [ ] Documentation is up-to-date
- [x] `CHANGES.md` is up-to-date

### CODE REVIEW

**Comment by mjambon** [comment]
Basic support for TypeScript was implemented and tested. I'll write the documentation in another PR. This PR is ready for merging.

File: .circleci/setup-system [hunk 1]
```diff
@@ -13,11 +13,17 @@ sudo apt-get update
...
   default-jdk \
+  nodejs \
+  npm \
   python3 \
   python3-pip \
   python-is-python3 \
   scala
 
+# For atdts
+sudo npm install -g typescript
```
Comment (Khady) [inline .circleci/setup-system:24]: it's a bit suspicious to have it install globally. I think that it should be a dev depedency in a package.json and then have a `script` field in this same package.json that would be like

```
"scripts": {
  "tsc": "tsc"
}
```

File: atdts/src/lib/Codegen.ml [hunk 2]
```diff
@@ -0,0 +1,1150 @@
...
+    | 'a'..'z' as c when !start_word ->
+        Buffer.add_char buf (Char.uppercase_ascii c);
+        start_word := false
+    | c ->
+        Buffer.add_char buf c;
+        start_word := false
+  done;
+  Buffer.contents buf
+
+(* Translate a preferred variable name into an available Python identifier. *)
```
Comment (Khady) [inline atdts/src/lib/Codegen.ml:62]:
```suggestion
(* Translate a preferred variable name into an available TypeScript identifier. *)
```

File: atdts/src/lib/Codegen.mli [hunk 3]
```diff
@@ -0,0 +1,7 @@
+(*
+   Python code generation for JSON support (no biniou support)
```
Comment (Khady) [inline atdts/src/lib/Codegen.mli:2]:
```suggestion
   TypeScript code generation for JSON support (no biniou support)
```

File: atdts/src/lib/Codegen.mli [hunk 4]
```diff
@@ -0,0 +1,7 @@
...
+*)
+
+(** Take ATD type definitions and translate them to Python, writing
+    them out to a file which should have the '.py' extension. *)
```
Comment (Khady) [inline atdts/src/lib/Codegen.mli:6]:
```suggestion
(** Take ATD type definitions and translate them to TypeScript, writing
    them out to a file which should have the '.ts' extension. *)
```

**Review by Khady — APPROVED** [review]
You are using native typescript json decoding without relying on an external runtime as it was done for bucklescript?

There's a decent amount of duplicated code between tools now. But can be solved another time.

I didn't look for bugs or really checked the runtime

**Comment by mjambon** [comment]
Sorry about all those mentions of python where it should be typescript. I should have fixed them before submitting the PR.

**Comment by mjambon** [comment]
> You are using native typescript json decoding without relying on an external runtime as it was done for bucklescript?

The runtime is included in each generated `.ts` file. It's about [300 lines](https://github.com/ahrefs/atd/blob/52ecbd530f5626bf1dc8b8cf2f767bea6aecf702/atdts/test/ts-expected/everything.ts#L143-L475). We could have it in a runtime library but it's more complicated.

> There's a decent amount of duplicated code between tools now. But can be solved another time.

Yes

**Comment by Khady** [comment]
> JSON parsing and printing is meant to be done by the [built-in JSON library](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON) (string <-> any). The code we generate validates and converts between the any value (JSON AST) and a typed value. A few things such as field names being renamed or the representation of variants change in the conversion, so it's not a pure validation. I'll try to explain this a bit in the documentation.

Sounds good. The goal in the bucklescript implementation was also to use the built-in JSON lib for good perf. Embedding the runtime makes it easier to use.

**Comment by Khady** [comment]
CHANGES.md is ticked but is not updated in this PR

**Comment by mjambon** [comment]
I changed the typescript/javascript setup to install dependencies locally + switched to strict type checking mode.

Merging!

File: .circleci/setup-system (hunk 1 above)
Comment (mjambon) [reply .circleci/setup-system]: done.

**Pull request merged**